modules/javafx.graphics/src/main/java/javafx/scene/image/PixelBuffer.java

```diff
@@ -0,0 +1,194 @@
+        this.width = width;
+        this.height = height;
+        this.pixelFormat = pixelFormat;
+    }
+
+    /**
+     * Returns the {@code java.nio.Buffer} of this {@code PixelBuffer}.
+     *
+     * @return the {@code java.nio.Buffer} of this {@code PixelBuffer}.
+     */
+    public T getBuffer() {
+        return buffer;
+    }
+
+    /**
+     * Returns the width dimension of {@code PixelBuffer}.
+     *
+     * @return the width dimension of this {@code PixelBuffer}.
+     */
+    public int getWidth() {
+        return width;
+    }
+
+    /**
+     * Returns the height dimension of {@code PixelBuffer}.
+     *
+     * @return the height dimension of this {@code PixelBuffer}.
+     */
+    public int getHeight() {
+        return height;
+    }
+
+    /**
+     * Returns the {@code PixelFormat} of the pixels in the {@code java.nio.Buffer}.
+     *
+     * @return the {@code PixelFormat} of the pixels in the {@code java.nio.Buffer}.
+     */
+    public PixelFormat getPixelFormat() {
+        return pixelFormat;
+    }
+
+    /**
+     * Updates the pixel data in {@code java.nio.Buffer} using provided {@link javafx.util.Callback}.
+     * This method must be called on the JavaFX Application thread.
+     * The {@code callback} method should update the {@code java.nio.Buffer} and
+     * return a {@code Rectangle2D} which encloses the modified region, or
+     * {@code null} to indicate that entire buffer is dirty.
+     * Pseudo-code how to use this method:
+     * <pre>{@code
+     * Callback<PixelBuffer, Rectangle2D> callback = pixelBuffer -> {
+     *     ByteBuffer buf = ((PixelBuffer<ByteBuffer>) pixelBuffer).getBuffer();
+     *     // Update the buffer.
+     *     return new Rectangle2D(x, y, width, height);
+     * };
+     * pixelBuffer.updateBuffer(callback);
+     * }</pre>
+     *
+     * @param callback the {@link javafx.util.Callback} method which updates the {@code java.nio.Buffer}.
+     * @throws IllegalStateException if this method is not being called on the JavaFX Application Thread.
+     * @throws NullPointerException  if {@code callback} is {@code null}.
+     **/
+    public void updateBuffer(Callback<PixelBuffer, Rectangle2D> callback) {
+        Objects.requireNonNull(callback, "callback must be specified.");
+        Toolkit.getToolkit().checkFxUserThread();
+        Rectangle2D rect2D = callback.call(this);
+        Rectangle rect = null;
+        if (rect2D != null) {
+            rect = new Rectangle((int) rect2D.getMinX(), (int) rect2D.getMinY(),
+                    (int) rect2D.getWidth(), (int) rect2D.getHeight());
+        }
+        bufferDirty(rect);
+    }
+
+    private void bufferDirty(Rectangle rect) {
+        Iterator<WeakReference<WritableImage>> iter = imageRefs.iterator();
+        while (iter.hasNext()) {
+            WeakReference<WritableImage> imageRef = iter.next();
+            if (imageRef.get() != null) {
+                imageRef.get().bufferDirty(rect);
```

Comment:
There is a very small, but non-zero chance that the image could be garbage-collected between these two calls to `WeakReference::get`, which would lead to an NPE. A safer pattern is to store `imageRef.get()` in a temporary (can be final) variable that you then test and use.